File: app/server/notify/channels/serverChan.ts
/**
 * Server酱 推送渠道
 */
import Logger from '../../utils/logger';
import { NotifyChannel, NotifyResult } from '../types';
import { $ } from '../httpClient';
import { hasConfig, getConfig } from '../config';

const logger = Logger.child({ channel: 'serverChan' });

export const serverChanChannel: NotifyChannel = {
    name: 'serverChan',
    enabled: false,

    async send(text: string, desp: string): Promise<NotifyResult> {
        return new Promise((resolve) => {
            if (!hasConfig('PUSH_KEY')) {
                resolve({ success: false, message: 'PUSH_KEY 未配置' });
                return;
            }

            const PUSH_KEY = getConfig('PUSH_KEY') || '';

            // 微信server酱推送通知一个\n不会换行，需要两个\n才能换行
            const formattedDesp = desp.replace(/[\n\r]/g, '\n\n');

            // 判断是 Turbo 版还是旧版
            const matchResult = PUSH_KEY.match(/^sctp(\d+)t/i);
            const url = matchResult && matchResult[1]
                ? `https://${matchResult[1]}.push.ft07.com/send/${PUSH_KEY}.send`
                : `https://sctapi.ftqq.com/${PUSH_KEY}.send`;

            const body = `text=${encodeURIComponent(text)}&desp=${encodeURIComponent(formattedDesp)}`;

            $.post({
                url,
                body,
                headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
            }, (err, _resp, data) => {
                if (err) {
                    logger.error({ err }, 'Server酱 发送失败');
                    resolve({ success: false, error: err as Error });
                } else {
                    // server酱和Server酱·Turbo版的返回json格式不太一样
                    const result = data as { errno?: number; data?: { errno?: number }; errmsg?: string };
                    if (result?.errno === 0 || result?.data?.errno === 0) {
                        logger.info('Server酱 发送成功');
                        resolve({ success: true, message: '发送成功' });
                    } else if (result?.errno === 1024) {
                        // 一分钟内发送相同的内容会触发
                        logger.warn({ msg: result.errmsg }, 'Server酱 发送频率限制');
                        resolve({ success: false, message: result.errmsg || '发送频率限制' });
                    } else {
                        logger.warn({ body: data }, 'Server酱 发送失败');
                        resolve({ success: false, message: '发送失败' });
                    }
                }
            });
        });
    }
};

// 检查是否启用
if (hasConfig('PUSH_KEY')) {
    serverChanChannel.enabled = true;
}

export default serverChanChannel;

File: app/server/internal/notify/channels/serverchan.go
package channels

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"net/url"
	"regexp"
	"strings"

	"github.com/sushazhi/fnos-logmanager/internal/notify"
)

type ServerChan struct {
	enabled bool
}

func (c *ServerChan) Name() string            { return "serverChan" }
func (c *ServerChan) Enabled() bool           { return c.enabled }
func (c *ServerChan) SetEnabled(enabled bool) { c.enabled = enabled }

func (c *ServerChan) Send(text, desp string) notify.NotifyResult {
	if !notify.HasConfig("PUSH_KEY") {
		return notify.NotifyResult{Success: false, Message: "PUSH_KEY 未配置"}
	}

	pushKey := notify.GetConfig("PUSH_KEY")

	// Replace single newlines with double newlines (ServerChan requires double newlines for proper line breaks)
	formattedDesp := strings.NewReplacer("\r\n", "\n\n", "\r", "\n\n", "\n", "\n\n").Replace(desp)

	// Determine if Turbo version or old version
	re := regexp.MustCompile(`(?i)^sctp(\d+)t`)
	matches := re.FindStringSubmatch(pushKey)
	var pushURL string
	if len(matches) > 1 && matches[1] != "" {
		pushURL = fmt.Sprintf("https://%s.push.ft07.com/send/%s.send", matches[1], pushKey)
	} else {
		pushURL = fmt.Sprintf("https://sctapi.ftqq.com/%s.send", pushKey)
	}

	body := fmt.Sprintf("text=%s&desp=%s", url.QueryEscape(text), url.QueryEscape(formattedDesp))

	resp, err := notify.HTTPPost(pushURL, notify.HttpRequestOptions{
		Body:    body,
		Headers: map[string]string{"Content-Type": "application/x-www-form-urlencoded"},
	})
	if err != nil {
		slog.Error("Server酱 发送失败", "err", err)
		return notify.NotifyResult{Success: false, Message: fmt.Sprintf("发送失败: %v", err), Err: err}
	}

	// Try top-level errno first, then nested data.errno
	var topResult struct {
		Errno  float64 `json:"errno"`
		Errmsg string  `json:"errmsg"`
		Data   *struct {
			Errno float64 `json:"errno"`
		} `json:"data"`
	}
	if err := json.Unmarshal([]byte(resp.Body), &topResult); err != nil {
		slog.Error("Server酱 解析响应失败", "err", err)
		return notify.NotifyResult{Success: false, Message: "解析响应失败", Err: err}
	}

	if topResult.Errno == 0 || (topResult.Data != nil && topResult.Data.Errno == 0) {
		slog.Info("Server酱 发送成功")
		return notify.NotifyResult{Success: true, Message: "发送成功"}
	}

	if topResult.Errno == 1024 {
		slog.Warn("Server酱 发送频率限制", "msg", topResult.Errmsg)
		return notify.NotifyResult{Success: false, Message: topResult.Errmsg}
	}

	slog.Warn("Server酱 发送失败", "body", resp.Body)
	return notify.NotifyResult{Success: false, Message: "发送失败"}
}

func init() {
	ch := &ServerChan{}
	if notify.HasConfig("PUSH_KEY") {
		ch.SetEnabled(true)
	}
	notify.Registry.Register(ch)
}
